### ISSUE TITLE
[WIP] Environment

### ISSUE THREAD

**Issue by regexident** [regexident/rspec#1]
@mackwic: I pushed an experimental branch ([environment](https://github.com/regexident/rspec/tree/environment)) with support for supplying environments to unit tests and their nested contexts, as declared in https://github.com/mackwic/rspec/issues/7#issuecomment-270071406.

There are quite a few rough edges still, and even more commented-out unit tests (:innocent:), but the overall design seems to be working. I figured it might be a good time now to start discussing whether this was a path we want to take, etc.

The refactoring is at a very early thinking-out-loud stage and nowhere near what'd be necessary for an actual PR. Thus I opted to push to my temporary fork and move the (early!) discussion. As soon as the noise-to-signal ratio settles down I'd then create an actual PR on upstream with proper technical discussion. Consider this more a high-noise brainstorming sandbox.

---

The API now consists of these functions for creating suites/contexts/examples:

- **Suite**: `describe()`, `suite()`, `given()`
- **Contexts**: `describe()`, `specify()`, `context()`, `given()`, `when()`
- **Examples**: `it()`, `example()`, `then()`

In order to reduce redundancy and streamline the API I'd propose reducing them to these:

- **Suite**: `describe()`, `suite()`, `given()`
- **Contexts**: `specify()`, `context()`, `when()`
- **Examples**: `it()`, `example()`, `then()`

Which would map to three possible syntactic styles:

1. `describe()` -> `specify()` -> `it()`

```rust
let env = ();
let mut runner = rspec::describe("suite name", env, |ctx| {
    ctx.it("example name", |env| {});
    ctx.specify("context name", |ctx| {
        ctx.it("example name", |env| {});
    });
});
```

2. `suite()` -> `context()` -> `example()`

```rust
let env = ();
let mut runner = rspec::suite("suite name", env, |ctx| {
    ctx.example("example name", |env| {});

    ctx.context("context name", |ctx| {
        ctx.example("example name", |env| {});
    });
});
````

3. `given()` -> `when()` -> `then()`

```rust
let env = ();
let mut runner = rspec::given("suite name", env, |ctx| {
    ctx.then("example name", |env| {});
    ctx.when("context name", |ctx| {
        ctx.then("example name", |env| {});
    });
});
```

---

Further more I modified the `Simple` formatter to make it match the output of Rust's native unit tests (with the addition of nesting).

To get an idea of how such an output might look like (including console colors), run the ['environment'](https://github.com/mackwic/rspec/blob/environment/examples/simple.rs) example:

```bash
$ cargo run --example simple
```

**Comment by regexident** [comment]
Any comments/critique, @mackwic?

If not I'd proceed further on this path, fixing the unit tests and documentation in preparation of a proper PR.

**Comment by mackwic** [comment]
Yes I saw it, I wanted to test how practical it was with the `HashSet` example. How much did you test it ?

**Comment by mackwic** [comment]
Sorry for the delay. I am ok with your proposal after some thoughts. We must use macros in the end anyway (in order to add line numbers and context).

Would you like some help ?

**Comment by regexident** [comment]
> Sorry for the delay. I am ok with your proposal after some thoughts.

Cool. I'll make sure to clean up the mess then and post a proper PR on upstream.

> We must use macros in the end anyway (in order to add line numbers and context).

Yeah, good point.

> Would you like some help?

At some point, maybe, but right now there's so much stuff I had to break (basically all the unit tests) to be able to experiment and brainstorm, that I'll have to go through the code base line by line and clean up the mess I made.
I'll (try to) do this on this WE and then we can do an assessment of what's missing or needs fixing/changes, maybe?
